### DISCUSSION TITLE
Bootstrap cerbos with storage type: "database" ?

### DISCUSSION THREAD

**Discussion by David3Ar** [cerbos/cerbos#1610] [Q&A] [answered]
Hey, 

we plan to use cerbos with database as storage type and we are wondering if there is some easy way to start cerbos with an initial data migration? (f.e. json policies that where red and persisted to mysql database at very first startup)

We want to run cerbos inside a private network and we will not be able to access the admin api at the very first beginning of our implementation. Are there any recommendations ? 

Best,
Davide

**Comment by charithe** [comment, marked as answer]
You could start an instance of Cerbos on a dev machine with a local MySQL instance, use [`cerbosctl put`](https://docs.cerbos.dev/cerbos/latest/cli/cerbosctl.html#put) to load up your initial set of policies, and use a tool like `mysqldump` to export the data as a SQL script that you can then use to seed your production database.

If you don't have access to the database either, you could do a slight variation of this by creating a bootstrap script that uses `cerbosctl put` to load the policies into Cerbos once it has started up inside the private network. The invocation of `cerbosctl put` is idempotent so you don't have to worry about running it more than once.

**Reply by David3Ar** [reply]
Great! Thanks!

**Reply by David3Ar** [reply]
@charithe it would be very helpful if we could bootstrap cerbos with mounted yaml policies that will be directly imported into database. 

At least it would fit well in our concept, while we are waiting that AdminAPI will be accessible from JavaSDK.

**Comment by G-Salviati** [comment]
Hello,
I'm trying to deploy cerbos in a k8s cluster with the helm chart.
I'm currently using the sqlite3 driver but i'm pretty sure i'm going to use the postgres or mysql driver.

I'm trying to migrate some initial policies, schemas and tests to the cerbos pod through the cerbosctl docker solution. 
Specifically I'm having troubles finding a way to use the put policy command with the recursive flag due to my policies folder structure. I'm following the [best practices repository structure](https://docs.cerbos.dev/cerbos/latest/policies/best_practices). So my policies folder has 3 layers of subfolders (e.g: resource_policies -> hr -> testdata) and if I try to put the whole policies folder it says it is an empty set of policies. 

The only way to make it work is to have a folder full of yaml files only, I guess. 
Is there an easy way to achieve my initial migration? 

Briefly, my aim is to migrate the policies folder (following the best practices I've linked and including schemas, tests, derived roles) at the bootstrap of the pod or immediately afterwards. 

Thank you!

**Reply by charithe** [reply]
Hey, the `--recursive` or `-R` option to `cerbosctl put` should traverse subdirectories. I assume you're using a volume to load the files. Is it possible that the directory structure on the volume is different from what you think it should be? Another problem could be the volume permissions.

Cerbosctl now supports zip/tar files as well. So, you could also create a ConfigMap or Secret containing the zipped contents of your policy repository and use `cerbosctl` to load that zip file. We do something similar in our tests.

```sh
tar -czf "$TEMP_FILE" -C "${SOURCE_DIR}/policies" .
kubectl create secret generic cerbos-policy-tarball \
    --from-file="policies.tgz=${TEMP_FILE}" \
    --dry-run=client \
    -o yaml | kubectl apply -f -
```

You can then mount that secret as a volume and run `cerbosctl`:

```
cerbosctl put policy /volume/mount/path/policies.tgz
```

**Reply by G-Salviati** [reply]
Hello,
thank you for you help! I was able to load the policies through the cerbosctl docker with a command similar to the following one 
```
docker run --network="host" -it -v ./policies:/policies ghcr.io/cerbos/cerbosctl:0.41.0 --server=localhost:7777 --username=<USERNAME>--password=<PASSWORD> --plaintext --insecure put policy -R ./policies
```

Anyway, I have new troubles: 
1) Even if the engine recognizes the resource policies stored in the "policies/resource_policies" folder and the derived roles in "policies/derived_roles" it can't recognize the json schema in "policies/_schemas". When I load the policies folder there's no error printed in the logs. If I use the GUI and the "get /admin/schemas" no schema is returned. If I test the resource_policies through the "/api/check/resources" no request validation is done. I've tried to add "schema.enforcment: warn" in the values.yaml for the helm chart and nothing has changed. I wanted to enter in the pod's filesystem but no shell is included in the docker image since it's based from scratch (and honestly I don't know how to add a shell).

2) When I upload the policies folder I have no feedback regarding the tests. Do the engine run tests automatically? The tests are located in the same folders of the resource policies as showed in the best practices page. 

Note that I've tried everything in the cerbos playground and it works. Schema and tests are recognized and work as inteded.

Thank you!

**Reply by charithe** [reply]
The schemas need to be added separately with `cerbosctl put schema`. I realise it's a bit confusing to have to use two commands instead of one. We are planning to fix that soon. 

Cerbos doesn't run tests when there are new policies added to the store. It expects a GitOps workflow where the tests are run by an external process such as a CI pipeline and only promoted to production (i.e. added to the running Cerbos instance) when the tests pass.
